Promote egress gateway to stable

jibi opened the pull request:
See individual commits.

jibi committed:
k8s: add egressGateway property to CiliumEgressNATPolicy

This commit adds a new egressGateway property to the
CiliumEgressNATPolicy object, which improves the way the egress gateway
node is selected and configured.

Before this change a CiliumEgressNATPolicy would only allow to specify a
an egressSourceIP, which was used both as:

* IP to masquerade traffic leaving the cluster
* IP to identify which node was acting as gateway for the policy

Now, with the egressGateway property it's possible to explicitly
identify which node should act as gateway by using a nodeSelector based
on node labels. In this way, workers node will forward traffic to the
egress gateway node by using its internal node IP.

Moreover, the new property extends the way the egress IP can be
configured, as a user can either:

* specify the egress IP that should be used to masquerade traffic
* specify the interface that should be used to masquerade traffic: in
  this case the first IPv4 assigned to that interface will be used as
  egress IP
* not specify anything: in this case the first IPv4 assigned to the
  interface for the default route will be used

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
egressgw: manager: add support for new egressGateway policy's property

This commit adds support to the manager for the new egressGateway
property, which improves the way the egress gateway node is selected and
configured.

At a high level these changes can be split into 3 main areas:

* manager subscribing to k8s CiliumNode events, in order to track which
  labels are assigned to which node
* manager being able to parse the new egressGateway policy field
* manager keeping additional state (gatewayConfig) for each policy,
  which is updated during each reconciliation run and is used to store,
  in addition to the configuration specified in the policy, additional
  information that can only be derived from the running cluster, such as
  the internal node IP of the gateway node or the interface index used
  to SNAT traffic

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
egressgw: manager: make mutex an embedded field

given that there's only a single mutex protecting the entire struct, we
can embed it and call the Lock/Unlock methods directly on the Manager
object.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
egressgw: add support for ENI IP rules/routes

This commit introduces a new flag, --install-egress-gateway-routes
which if enabled will install on the gateway nodes the required IP rules
and routes to steer gateway traffic to the correct ENI interface

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
cilium: remove bpf egress {get, update, delete} commands

These commands were useful for debugging purposes, but now it doesn't
really make sense to configure a policy manually through the CLI

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

tklauser commented on pkg/egressgateway/endpoint.go:
```diff
@@ -7,10 +7,10 @@ import (
 	k8sTypes "github.com/cilium/cilium/pkg/k8s/types"
 	"github.com/cilium/cilium/pkg/labels"
+
+	"k8s.io/apimachinery/pkg/types"
```
Nit: I think the order of the import groups should stay as is, i.e. 1) standard library, 2) third-party and 3) local packages. See #18357 for context.

tklauser commented on pkg/egressgateway/manager.go:
```diff
@@ -4,16 +4,19 @@
+	nodeTypes "github.com/cilium/cilium/pkg/node/types"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/types"
```
Same comment about import group ordering applies here.

tklauser commented on pkg/egressgateway/manager.go:
```diff
@@ -4,16 +4,21 @@
+
+	"github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+	"k8s.io/apimachinery/pkg/types"
```
Same comment about import group order applies here.

tklauser commented on pkg/egressgateway/policy.go:
```diff
@@ -7,18 +7,51 @@ import (
 	"github.com/cilium/cilium/pkg/policy/api"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/types"
```
Same comment about import group order applies here.

tklauser reviewed (commented):
Some nit about import group ordering inline. Not sure why golangci-lint is not complaining about these given https://github.com/cilium/cilium/blob/f1b52db3374f18d6a0da695198bd8388fe4b3dae/.golangci.yaml#L54-L55

julianwiedmann commented on pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml:
```diff
@@ -166,6 +166,83 @@ spec:
+                      fields is specified, the policy will use the first IPv4 assigned
+                      to the interface with the default route."
+                    type: string
+                  nodeSelector:
```
Should the `nodeSelector` be marked as a required property? Or is a config with only `egressIP` valid as well? 
(effectively that would be the same as using `egressSourceIP` ...?)

jibi replied:
It should be marked as required :+1: (I thought it was already by default but apparently it's not)

julianwiedmann commented on pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml:
```diff
@@ -176,7 +253,6 @@ spec:
             required:
             - destinationCIDRs
             - egress
-            - egressSourceIP
```
Should we prevent users from building a policy with no `egressSourceIP` and no `egressGateway`? Using `oneOf` maybe?

jibi replied:
I'm already enforcing that in the logic that parses the policy:
https://github.com/cilium/cilium/blob/c2e37d36f3ba58665fb735d284e93c03c7bf275b/pkg/egressgateway/policy.go#L227-L232 

~but I should probably do it using the `oneOf` annotation in the type definition :+1:~
looks like there's been some discussion around that (https://github.com/kubernetes-sigs/controller-tools/issues/461) but it's not been implemented yet :disappointed:

julianwiedmann replied:
Looks like we have support for `oneOf` in [our fork](https://github.com/cilium/controller-tools) though :grin:  (commit ad007ee307a33bc6d45aafd383463c6dafd6e746).

julianwiedmann commented on Documentation/cmdref/cilium_bpf_egress.md:
```diff
@@ -21,8 +21,5 @@ Manage the egress routing rules
-* [cilium bpf egress get](../cilium_bpf_egress_get)	 - Get egress entries
 * [cilium bpf egress list](../cilium_bpf_egress_list)	 - List egress policy entries
-* [cilium bpf egress update](../cilium_bpf_egress_update)	 - Update egress entries
 
```
> but now it doesn't really make sense to configure a policy manually through the CLI

because the policy is getting too complex for a CLI-based configuration?

jibi replied:
> because the policy is getting too complex for a CLI-based configuration?

not really, in the end every policy can be distilled down to a list of `(source IP, dest CIDR, gateway IP, egress IP)` entries in the egress policy bpf map (and those commands were used exactly for this, to interact just with the bpf map).

The problem is that after the initial implementation we changed the way the manager reacts to events. Now it:
* keeps an internal state of all the objects it is interested in (nodes, endpoints, egress policies)
* whenever there's a new event (e.g. a new endpoint is added or a policy is updated) it updates its internal state to reflect that
* finally, it [reconciles](https://github.com/cilium/cilium/blob/4b1cb31e0c32e956ebd673a847a02cba63700779/pkg/egressgateway/manager.go#L371-L391) the state of the system with its internal state (so it distills the list of nodes, ep and policies down to a list of entries for the bpf map, adds the ones that are missing and removes the ones that should not be present anymore)

Because of this we should not allow a user to manually interact with the map entries, since whenever there's a new event (for example a new endpoint added) the manager will run its reconciliation logic, and will just delete (or add back) the entries that the user manually created (or deleted)

julianwiedmann commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -74,7 +78,10 @@ func (k *K8sWatcher) ciliumNodeInit(ciliumNPClient *k8s.K8sCiliumClient, asyncCo
 							}
 							n := nodeTypes.ParseCiliumNode(ciliumNode)
 							errs := k.CiliumNodeChain.OnUpdateCiliumNode(oldCN, ciliumNode, swgNodes)
-							if isLocal {
```
This change from `isLocal` to `n.IsLocal()` feels unrelated ... ?

pchaigno commented on pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml:
```diff
@@ -40,6 +40,15 @@ spec:
+            - properties:
+                egressGateway: {}
+              required:
+              - egressGateway
```
Could you add an example in `examples/kubernetes-egress-gateway/`? Mostly interested in that to understand how the structure is changing, but probably also useful to have for documentation purposes.

jibi replied:
I was going to add it as part of the documentation changes, but I can do it now :+1:

pchaigno commented on pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml:
```diff
@@ -166,6 +175,85 @@ spec:
+                      redirected to node with IP 192.168.1.100 and SNAT’ed with IP
+                      address 192.168.1.100. \n When none of the Interface or EgressIP
+                      fields is specified, the policy will use the first IPv4 assigned
+                      to the interface with the default route."
```
If I understand correctly, `egressGateway.egressIP` is equivalent to `egressSourceIP`? Are we planning to deprecate `egressSourceIP` in a future version of the CRD?

jibi replied:
I was considering deprecating it, but I wanted to get some feedback first before actually marking it as such

pchaigno commented on pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml:
```diff
@@ -166,6 +175,85 @@ spec:
+                properties:
+                  egressIP:
+                    description: "EgressIP is a source IP address that the egress
+                      traffic is redirected to and SNATed with. \n Example: When it
```
What happens if the `egressIP` and the `nodeSelector` (and/or the `interface`) don't point to the same IP or even to the same node?

jibi replied:
Sorry, the comment is wrong here :disappointed: `egressIP` is just the IP used to masquerade traffic, but we'll use the node internal IP to forward traffic to the gateway node (given that we have a `NodeSelector` that identifies it and so we can use that to get its internal IP). I'll fix the description

pchaigno commented on pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml:
```diff
@@ -166,6 +175,85 @@ spec:
+                    description: This is a label selector which selects the node that
+                      should act as egress gateway for the given policy. In case multiple
+                      nodes are selected, only the first one in the lexicograph ordering
+                      will be used. This field follows standard label selector semantics.
```
> lexicograph ordering

lexical ordering?

Worth stating that it's over node names?

pchaigno commented on Documentation/gettingstarted/egress-gateway.rst:
```diff
@@ -7,11 +7,9 @@
 **********************
 
-.. include:: ../beta.rst
-
```
Not sure we want to backport this change to v1.11. Are we about to say the egress gateway is now stable in v1.11 as well? Then, what are we announcing in the v1.12 release blog post? We usually (e.g., host firewall, IPsec) mark the feature stable only in the documentation for the release where it became stable.

pchaigno replied:
Also, shouldn't this change be in its own commit as the very least?

jibi replied:
This whole PR is not going to be backported to v1.11 (and this change slipped through but should go with the rest of the docs changes). I'll remove it for now

pchaigno commented on pkg/egressgateway/manager.go:
```diff
@@ -159,65 +168,75 @@ func (manager *Manager) OnDeleteEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
+	manager.nodes = []nodeTypes.Node{}
+	for _, n := range manager.nodeDataStore {
+		manager.nodes = append(manager.nodes, n)
+	}
```
I don't understand the need for `manager.nodeDataStore`. Couldn't it be replaced by `manager.nodes` directly?

jibi replied:
It's not required as we could do everything with just the `manager.nodes` slice, but it simplifies the code a bit whenever we add or remove a node (as we just add/remove nodes to the map and then sort the list we derive from that map)

pchaigno commented on pkg/egressgateway/manager.go:
```diff
@@ -159,65 +168,75 @@ func (manager *Manager) OnDeleteEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
+
+	for _, policyConfig := range manager.policyConfigs {
+		policyConfig.forEachEndpointAndDestination(manager.epDataStore, addEgressRule)
+	}
```
I would have been nice to move this bit of refactoring to its own commit.

jibi replied:
Done :+1:

pchaigno commented on pkg/egressgateway/manager.go:
```diff
@@ -227,26 +246,21 @@ func (manager *Manager) removeUnusedEgressRules() {
-							continue nextPolicyKey
-						}
-					}
-				}
```
Ditto here regarding refactoring changes.

pchaigno commented on pkg/egressgateway/net.go:
```diff
@@ -0,0 +1,53 @@
+		}
+	}
+
+	return net.IPNet{}, 0, fmt.Errorf("no IPv4 address assigned to interface")
```
Nit:
```suggestion
	return net.IPNet{}, 0, fmt.Errorf("no IPv4 address assigned to interfaces")
```

nebril commented on pkg/egressgateway/manager.go:
```diff
@@ -281,6 +373,17 @@ func (manager *Manager) reconcile() {
+		shouldRetry := manager.addMissingIpRulesAndRoutes()
+		manager.removeUnusedIpRulesAndRoutes()
+
+		if shouldRetry {
```
Is one retry always enough?

jibi replied:
(synced offline)

nebril reviewed (approved):
Looks good overall, one question left inline, didn't review BPF bits.

pchaigno commented on pkg/egressgateway/manager.go:
```diff
@@ -159,65 +168,75 @@ func (manager *Manager) OnDeleteEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
+
+	for _, policyConfig := range manager.policyConfigs {
+		policyConfig.forEachEndpointAndDestination(manager.epDataStore, addEgressRule)
+	}
```
This is also only refactor AFAICS, right?

pchaigno commented on pkg/egressgateway/policy.go:
```diff
@@ -62,6 +222,29 @@ func ParsePolicy(cenp *v2alpha1.CiliumEgressNATPolicy) (*PolicyConfig, error) {
+		return nil, fmt.Errorf("CiliumEgressNATPolicy cannot have both EgressSourceIP and EgressGateway set")
+	case egressIP == nil && egressGateway == nil:
+		return nil, fmt.Errorf("CiliumEgressNATPolicy needs either EgressSourceIP or EgressGateway set")
+	}
```
Do we need this? IIUC, we already have a check in the CRD schema.

jibi replied:
Yes, I just switched the CRD to use `OneOf` but forgot to remove this

pchaigno commented on pkg/egressgateway/policy.go:
```diff
@@ -62,6 +222,29 @@ func ParsePolicy(cenp *v2alpha1.CiliumEgressNATPolicy) (*PolicyConfig, error) {
+	if egressGateway != nil {
+		if egressGateway.Interface != "" && egressGateway.EgressIP != "" {
+			return nil, fmt.Errorf("CiliumsEgressNATPolicy's gateway configuration can't specify both an interface and an egress IP")
+		}
```
Couldn't we prevent this as part of the CRD schema?

jibi replied:
I don't think we can validate this in the CRD though as `OneOf` semantic is "exactly one", while in this case we allow the user to not specify anything (and just use the first IPv4 assigned to the interface with the default route)

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -1131,6 +1131,79 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
+			 */
+			if (is_cluster_destination(ip4, remote_ep->sec_label,
+						   remote_ep->tunnel_endpoint))
+				goto skip_egress_gateway;
```
AFAIU, at this point we know the packet isn't going to a local endpoint so we can remove one map lookup (`lookup_ip4_endpoint`) from `is_cluster_destination` (probably needs a big comment for the function though).

jibi replied:
The problem is that that function is used also in `snat_v4_needed()`, I need to check if it's possible to refactor it

jibi replied:
Let's recap what's going on:
* we are calling `remote_ep = lookup_ip4_remote_endpoint(ip4->daddr)`, which looks up the ep in the ipcache map
* we are passing `remote_ep->sec_label` and `remote_ep->tunnel_endpoint` to `is_cluster_destination()`
* in `is_cluster_destination()` we are calling `lookup_ip4_endpoint(ip4)` (i.e. `__lookup_ip4_endpoint(ip4->daddr)`) which looks up the ep in the endpoint map

To my understanding (and after some tests/printks):
* `lookup_ip4_remote_endpoint()` will return a `struct remote_endpoint_info *` both for local and remote endpoints
* `lookup_ip4_endpoint()` will only return a `struct endpoint_info *` if the endpoint is local

so we could just pass `remote_ep` to `is_cluster_destination()` and get rid of the `lookup_ip4_endpoint(ip4)` check (if `remote_ep` is NULL it means that's not a local/remote ep, so the `lookup_ip4_endpoint(ip4)` case would be covered)

edit: `remote_ep` is never going to be NULL, so we would need to check the actual sec label (see below)

jibi replied:
Actually, could we simplify that logic even more and just check `remote_ep->sec_label`? 

```c
static __always_inline bool
is_cluster_destination(struct remote_endpoint_info *remote_ep)
{
    if (!remote_ep)
        return false; // just to keep the verifier happy as the ipcache is always going to return at least the `0.0.0.0/0` entry
    if (remote_ep->sec_labels == WORLD_ID)
        return false

    return true
}
```

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -1131,6 +1131,79 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
+		 */
+		local_ep = __lookup_ip4_endpoint(egress_gw_policy->gateway_ip);
+		if (local_ep && (local_ep->flags & ENDPOINT_F_HOST))
+			goto skip_egress_gateway;
```
If the gateway is the local node, then `remote_ep->sec_label == HOST_ID`, no?

jibi replied:
`remote_ep` refers to the EP associated with the actual destination IP of the packet (while here we are looking up the gateway IP which should be a node internal IP)

pchaigno replied:
Oh, of course :facepalm: 

I still wonder if there's a more efficient way we could check this. Once we remove `egressIP` in the CRD can `gateway_ip` still be any IP address assigned to the gateway node or will it only be the InternalIP (that used for tunneling)?

jibi replied:
Once `egressSourceIP` is removed the manager should only assign internal IPs to the `gateway_ip` field in the policy map

pchaigno replied:
Ok, so when that happens, we should be able to avoid one map lookup here by instead checking if `egress_gw_policy->gateway_ip == INTERNAL_NODE_IP`.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -848,46 +848,6 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
 	}
 #endif /* ENABLE_HOST_FIREWALL && !ENABLE_ROUTING */
 
-#ifdef ENABLE_EGRESS_GATEWAY
```
Do we still have any egress gateway logic in `bpf_lxc.c`? If not, we can update `bpf/Makefile` and `bpf/complexity-tests/*/bpf_lxc.txt` a bit.

julianwiedmann commented on pkg/egressgateway/net.go:
```diff
@@ -0,0 +1,176 @@
+		return net.IPNet{}, 0, err
+	}
+
+	addrs, err := netlink.AddrList(dev, netlink.FAMILY_ALL)
```
Would we benefit from

```suggestion
	addrs, err := netlink.AddrList(dev, netlink.FAMILY_V4)
```

julianwiedmann commented on pkg/egressgateway/net.go:
```diff
@@ -0,0 +1,176 @@
+	}
+
+	for _, l := range links {
+		addrs, err := netlink.AddrList(l, netlink.FAMILY_ALL)
```
Also

```suggestion
		addrs, err := netlink.AddrList(l, netlink.FAMILY_V4)
```

?

julianwiedmann commented on pkg/egressgateway/policy.go:
```diff
@@ -44,6 +81,129 @@ func (config *PolicyConfig) selectsEndpoint(endpointInfo *endpointMetadata) bool
+			return fmt.Errorf("failed to retrieve interface with egress IP: %w", err)
+		}
+	default:
+		// If the gateway config doesn't specify any egress IP or interface, us
```
```suggestion
		// If the gateway config doesn't specify any egress IP or interface, use
```

julianwiedmann commented on pkg/egressgateway/policy.go:
```diff
@@ -10,15 +10,48 @@ import (
+// Some of these fields are derived from the running system as the policy may
+// specify only the egress IP (and so we need to figure out which interface has
+// that IP assigned to) or the interface (and in this case we need to find the
+// first IPv4 assigned to that).
```
I think we might want to point this out in the docs as well - any subsequent change to a node's egress interface config won't be reconciled with the policy.

brb commented on bpf/bpf_host.c:
```diff
@@ -1131,6 +1131,79 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
 
+#if defined(ENABLE_EGRESS_GATEWAY) && defined(ENABLE_IPV4)
+	{
+		void *data, *data_end;
```
I think this code block is big enough to be moved to a separate function.

Also, JFYI we are adding the WG blob too to the similar place - https://github.com/cilium/cilium/pull/19407/files. The egress blob should go before the WG blob, as the redirect for the encryption will be handled by  `encap_and_redirect_with_nodeid()`. Just that the egress blob should be able to handle `CTX_ACT_REDIRECT`.

qmonnet commented on pkg/datapath/linux/linux_defaults/linux_defaults.go:
```diff
@@ -57,6 +64,12 @@ const (
 	RulePriorityWireguard = 1
 
+	// RulePriorityEgressGateway is the priority used in IP routes added by the manager.
+	// This value was picked as it's lower then the ones used by Cilium
```
```suggestion
	// This value was picked as it's lower than the ones used by Cilium
```

jibi committed:
egressgw: policy: add forEachEndpointAndDestination and matches helpers

which allow to hide some complexity around explicitly iterating through
each policy's combination of source IPs, destination CIDRs and egressIP.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
egressgw: require L7 proxy to be disabled

When support for L7 policies is enabled, traffic is redirected to the
proxy before the logic in bpf_lxc can inspect it to determine if it
should be forwarded, resulting in no traffic being forwarded to the
egress gateway node.

This commit adds an additional check to make sure egressgw can't be
enabled with L7 policies.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

pchaigno reviewed (approved):
Thanks for the new refactoring commit! :rocket:

qmonnet reviewed (approved):
Sorry for the late review. Looks all good to me, nice work!

Just two (non-blocking) comments:
- Should the new agent option be documented in Documentation/operations/upgrade.rst (especially now that egress gateway is stable)?
- Maybe a matter of personal preferences, but I found it would be nice to have a bit more context/motivations in some commit logs that you left empty, so that we can understand better the rationale behind the changes when looking at the history.

julianwiedmann commented on pkg/egressgateway/policy.go:
```diff
@@ -54,13 +214,27 @@ func ParsePolicy(cenp *v2alpha1.CiliumEgressNATPolicy) (*PolicyConfig, error) {
+	var policyGwc *policyGatewayConfig
+	if egressGateway != nil {
+		if egressGateway.Interface != "" && egressGateway.EgressIP != "" {
+			return nil, fmt.Errorf("CiliumsEgressNATPolicy's gateway configuration can't specify both an interface and an egress IP")
```
Nitpick:
```suggestion
			return nil, fmt.Errorf("CiliumEgressNATPolicy's gateway configuration can't specify both an interface and an egress IP")
```

pchaigno commented on pkg/egressgateway/net.go:
```diff
@@ -0,0 +1,176 @@
+	out := make(net.IP, net.IPv4len)
+	copy(out, ip.IP.To4())
+	out = out.Mask(ip.Mask)
+	out[3] += 1
```
What's the logic here? If the egress IP is 10.0.1.45/32, then 10.0.1.46 will be selected as the gateway. Why?
